tests/run/i5077.scala

```diff
@@ -0,0 +1,32 @@
+trait Is[A]
+case object IsInt extends Is[Int]
+case object IsString extends Is[String]
+case class C[A](is: Is[A], value: A)
+
+@main
+def Test = {
+  val c_string: C[String] = C(IsString, "name")
+  val c_any: C[_] = c_string
+  val any: Any = c_string
+
+  // Case 1: Should give compile error, since IsInt is not compatible with C[String]
```

Comment:
This should get a changed comment since it does not give a compile error. It would be best if we explained why the behavior deviates from what is proposed in the issue.

Reply:
It would also be good to have a check file that verifies the exact runtime behavior of this code.